cmd/delete.go

```diff
@@ -0,0 +1,167 @@
+Root must be provided and can be a single file or directory.
+
+Dry-run mode is enabled by default, where a line will be output to state where a file would be deleted when running not in dry run mode.
+It is recommended to run using dry-run mode enabled, checking the logged output and then running with dry-run disabled using the same configuration flags.
+Use --dry-run=false to turn off dry-run mode, enabling deletes.
+
+When used on windows, deleting based on file ownership is unsupported and skip-owner-check should be used instead of directory-with-owner.
+`,
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			crawlConfig, err := createCrawlConfig(args[0], cmdCrawlFlags)
+			if err != nil {
+				return err
+			}
+
+			if len(directoriesWithOwners) > 0 && skipOwnerCheck {
+				return errors.New("--directory-with-owner and --skip-owner-check cannot be used together")
+			}
+			if len(directoriesWithOwners) == 0 && !skipOwnerCheck {
+				return errors.New("at least one --directory-with-owner value must be provided or --skip-owner-check must be set")
+			}
+
+			var ms []deleter.Matcher
+			for _, value := range directoriesWithOwners {
+				split := strings.Split(value, ":")
+				if len(split) != 2 {
+					return fmt.Errorf(`invalid directory-with-owner, must contain 2 colon-separated segments but got %q`, value)
+				}
+				expr, err := regexp.Compile(split[0])
+				if err != nil {
+					return fmt.Errorf("error compiling pattern for directory-with-owner %s: %w", value, err)
+				}
+				ms = append(ms, deleter.TemplatedOwner{
+					DirectoryExpression: expr,
+					OwnerTemplate:       split[1],
+				})
+			}
+
+			var minimumFindingsRequirement crawl.Finding
+			for _, value := range findingsMatches {
+				f, err := crawl.FindingOf(value)
+				if err != nil {
+					return err
+				}
+				minimumFindingsRequirement |= f
+			}
+
+			var detailedOutputWriter io.Writer
+			var enableTraceLogging bool
+			if !cmdCrawlFlags.disableDetailedFindings {
+				detailedOutputWriter = cmd.OutOrStdout()
+				enableTraceLogging = cmdCrawlFlags.enableTraceLogging
+			}
+
+			var filepathMatchFunc func(string) (bool, error)
+			if skipOwnerCheck {
+				filepathMatchFunc = func(string) (bool, error) {
+					return true, nil
+				}
+			} else {
+				filepathMatchFunc = (&deleter.FileOwnerMatchers{
+					Matchers:     ms,
+					ResolveOwner: snifferos.OwnerUsername,
+				}).Match
+			}
+
+			_, err = crawler.Crawl(cmd.Context(), crawlConfig, deleter.Deleter{
+				Logger: log.Logger{
+					OutputWriter:       detailedOutputWriter,
+					ErrorWriter:        cmd.OutOrStderr(),
+					EnableTraceLogging: enableTraceLogging,
+				},
+				FilepathMatch: filepathMatchFunc,
+				FindingMatch: func(finding crawl.Finding) bool {
+					return crawl.AllFindingsSatisfiedBy(minimumFindingsRequirement, finding)
+				},
+				DryRun: dryRun,
+				Delete: os.Remove,
+			}.Process, cmd.OutOrStdout(), cmd.OutOrStderr())
```

Comment:
Not a big deal to keep as-is, but does feel a bit awkward to be passing the `Process` function of the constructed struct -- feels like it may be better for the `Crawl` function to accept an interface instead and then define the functions on that interface.

However, that does have its own pros and cons as well. Probably fine to leave this as-is for this iteration, but wanted to mention for completeness.

Reply:
Let's keep as-is for the moment.
Happy to FLUP if necessary. 

I'm often put off creating a single method interface where the name forces implementations to have a vague method name. Perhaps I'm missing the huge pros of calling everything `Process` or something like that.

Reply:
Agree, happy to chat through this more later but think that we should defer for this PR.